include/pros/ext_adi.h

```diff
@@ -0,0 +1,720 @@
+ * instead.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports
+ * EADDRINUSE - The port is not configured as an analog input
+ *
+ * \param smart_port
+ *        The smart port number that the ADI Expander is on
+ * \param adi_port
+ *	      The ADI port number (from 1-8, 'a'-'h', 'A'-'H') to configure
+ *
+ * \return The difference of the sensor value from its calibrated default from
+ * -4095 to 4095
+ */
+int32_t ext_adi_analog_read_calibrated(uint8_t smart_port, uint8_t adi_port);
+
+/**
+ * Gets the 16 bit calibrated value of an analog input port.
+ *
+ * The adi_analog_calibrate() function must be run first. This is intended for
+ * integrated sensor values such as gyros and accelerometers to reduce drift due
+ * to round-off, and should not be used on a sensor such as a line tracker
+ * or potentiometer.
+ *
+ * The value returned actually has 16 bits of "precision", even though the ADC
+ * only reads 12 bits, so that error induced by the average value being between
+ * two values when integrated over time is trivial. Think of the value as the
+ * true value times 16.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports
+ * EADDRINUSE - The port is not configured as an analog input
+ *
+ * \param smart_port
+ *        The smart port number that the ADI Expander is on
+ * \param adi_port
+ *	      The ADI port number (from 1-8, 'a'-'h', 'A'-'H') to configure
+ *
+ * \return The difference of the sensor value from its calibrated default from
+ * -16384 to 16384
+ */
+int32_t ext_adi_analog_read_calibrated_HR(uint8_t smart_port, uint8_t adi_port);
+
+/**
+ * Gets the digital value (1 or 0) of a port configured as a digital input.
+ *
+ * If the port is configured as some other mode, the digital value which
+ * reflects the current state of the poprt is returned, which may or may not
+ * differ from the currently set value. The return value is undefined for ports
+ * configured as any mode other than a Digital Input.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports
+ * EADDRINUSE - The port is not configured as a digital input
+ *
+ * \param smart_port
+ *        The smart port number that the ADI Expander is on
+ * \param adi_port
+ *	      The ADI port number (from 1-8, 'a'-'h', 'A'-'H') to configure
+ *
+ * \return True if the pin is HIGH, or false if it is LOW
+ */
+int32_t ext_adi_digital_read(uint8_t smart_port, uint8_t adi_port);
+
+/**
+ * Gets a rising-edge case for a digital button press.
+ *
+ * This function is not thread-safe.
+ * Multiple tasks polling a single button may return different results under the
+ * same circumstances, so only one task should call this function for any given
```

Comment:
You may want to specify that this isn't thread safe with adi_digital_get_new_press as well.

Reply:
I don't think this is necessary.  The descriptions of both api’s already say that the function is not thread-safe with two threads referring to the same button at the same time.  I think that the current descriptions should be clear enough that the two functions do not work together.  Additionally, the two c api's are not intended to ever refer to the same device.  The external adi should only refer to ports 1-21 and the internal adi should only refer to port 22.  The external adi api can refer to the internal adi, but that is not its intended purpose for end-users.  Anyone who can understand that they can use the external adi api to refer to the internal should also be able to understand from the current warning that these two functions would conflict.